<a href="https://colab.research.google.com/github/FurqanAhmad019/24i-2534/blob/main/Lab01/24i_2534_DS_5A_Lab01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
# Task 1

df = pd.read_csv('/adult.csv')
df.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [ ]:
print(df.isnull().sum())
df.replace('?', np.nan, inplace=True)
count = df.isnull().sum()
print(count)

age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
income            0
dtype: int64
age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64


In [ ]:
# Task 2

rows_count, columns_count = df.shape
percentage = (count * 100 / rows_count)
print(percentage.sort_values(ascending = False))

occupation        5.660146
workclass         5.638647
native.country    1.790486
fnlwgt            0.000000
education         0.000000
education.num     0.000000
age               0.000000
marital.status    0.000000
relationship      0.000000
sex               0.000000
race              0.000000
capital.gain      0.000000
capital.loss      0.000000
hours.per.week    0.000000
income            0.000000
dtype: float64


In [ ]:
#Task 3

df['workclass'].replace(np.nan, "Unknown", inplace=True)
df['occupation'].replace(np.nan, "Unknown", inplace=True)
df.head()

#Null values in both workclass and occupation has been replaced with "Unknown".

/tmp/ipykernel_1828/3074269927.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['workclass'].replace(np.nan, "Unknown", inplace=True)
/tmp/ipykernel_1828/3074269927.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,Unknown,77053,HS-grad,9,Widowed,Unknown,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,Unknown,186061,Some-college,10,Widowed,Unknown,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [ ]:
#Task 4

print("Duplicated Rows: ", df.duplicated().sum())


Duplicated Rows:  24


In [ ]:
#Task 5

print("Unique values of Education: \n",df['education'].unique())
print("\nUnique values of Marital Status: \n", df['marital.status'].unique())
print("\nUnique values of Native Country: \n", df['native.country'].unique())



Unique values of Education: 
 ['HS-grad' 'Some-college' '7th-8th' '10th' 'Doctorate' 'Prof-school'
 'Bachelors' 'Masters' '11th' 'Assoc-acdm' 'Assoc-voc' '1st-4th' '5th-6th'
 '12th' '9th' 'Preschool']

Unique values of Marital Status: 
 ['Widowed' 'Divorced' 'Separated' 'Never-married' 'Married-civ-spouse'
 'Married-spouse-absent' 'Married-AF-spouse']

Unique values of Native Country: 
 ['United-States' nan 'Mexico' 'Greece' 'Vietnam' 'China' 'Taiwan' 'India'
 'Philippines' 'Trinadad&Tobago' 'Canada' 'South' 'Holand-Netherlands'
 'Puerto-Rico' 'Poland' 'Iran' 'England' 'Germany' 'Italy' 'Japan' 'Hong'
 'Honduras' 'Cuba' 'Ireland' 'Cambodia' 'Peru' 'Nicaragua'
 'Dominican-Republic' 'Haiti' 'El-Salvador' 'Hungary' 'Columbia'
 'Guatemala' 'Jamaica' 'Ecuador' 'France' 'Yugoslavia' 'Scotland'
 'Portugal' 'Laos' 'Thailand' 'Outlying-US(Guam-USVI-etc)']


In [ ]:
from numpy._core.defchararray import isnumeric
# Task 6

print(df.describe(include = [np.number]))
df.describe(include = "object")

                age        fnlwgt  education.num  capital.gain  capital.loss  \
count  32561.000000  3.256100e+04   32561.000000  32561.000000  32561.000000   
mean      38.581647  1.897784e+05      10.080679   1077.648844     87.303830   
std       13.640433  1.055500e+05       2.572720   7385.292085    402.960219   
min       17.000000  1.228500e+04       1.000000      0.000000      0.000000   
25%       28.000000  1.178270e+05       9.000000      0.000000      0.000000   
50%       37.000000  1.783560e+05      10.000000      0.000000      0.000000   
75%       48.000000  2.370510e+05      12.000000      0.000000      0.000000   
max       90.000000  1.484705e+06      16.000000  99999.000000   4356.000000   

       hours.per.week  
count    32561.000000  
mean        40.437456  
std         12.347429  
min          1.000000  
25%         40.000000  
50%         40.000000  
75%         45.000000  
max         99.000000  


,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32561,32561,32561,32561,32561,32561,32561,32561,32561
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720


In [ ]:
# Task 7

df.value_counts('occupation')
print("The most common occupation is: Prof-speciality")

male_count, female_count = df.value_counts('sex')
print("Female percentage: ", female_count * 100 / (male_count + female_count))
print("Male percentage: ", male_count * 100 / (male_count + female_count))

income_less_than_50k, income_greater_than_50k = df.value_counts('income')
print("Percentage of Income >= 50k: ", income_greater_than_50k * 100 / (income_greater_than_50k + income_less_than_50k))
print("Percentage of Income <= 50k: ", income_less_than_50k * 100 / (income_greater_than_50k + income_less_than_50k))

The most common occupation is: Prof-speciality
Female percentage:  33.07945087681582
Male percentage:  66.92054912318417
Percentage of Income >= 50k:  24.080955744602438
Percentage of Income <= 50k:  75.91904425539757


In [ ]:
# Task 8

df.groupby('education')['education.num'].unique()

,education.num
education,
10th,[6]
11th,[7]
12th,[8]
1st-4th,[2]
5th-6th,[3]
7th-8th,[4]
9th,[5]
Assoc-acdm,[12]
Assoc-voc,[11]


In [ ]:
# Task 9

import sqlite3

conn = sqlite3.connect(':memory:')

df.to_sql('adult_income', conn, if_exists='replace', index=False)

sql_query = "SELECT * FROM adult_income WHERE age > 30;"

df_sql = pd.read_sql_query(sql_query, conn)

print(f"Original DataFrame shape: {df.shape}")
print(f"SQL Query Result DataFrame shape (age > 30): {df_sql.shape}")

df_filtered_original = df[df['age'] > 30]
print(f"Original DataFrame filtered (age > 30) shape: {df_filtered_original.shape}")

if df_sql.shape == df_filtered_original.shape:
    print("\nShape comparison successful: The SQL extracted DataFrame matches the shape of the original DataFrame filtered by the same condition.")
else:
    print("\nShape comparison failed: Mismatch in DataFrame shapes.")

conn.close()

Original DataFrame shape: (32561, 15)
SQL Query Result DataFrame shape (age > 30): (21989, 15)
Original DataFrame filtered (age > 30) shape: (21989, 15)

Shape comparison successful: The SQL extracted DataFrame matches the shape of the original DataFrame filtered by the same condition.


# Task 10

There are 15 columns and 32561 rows. The dataset represents personal details of 15 * 32561 people including their education, marital status and income etc.

There are some missing values in the dataset, 24 duplicated rows but the data was consistent.

In the dataset, most of people are below 50k salary. The most common occupation is Prof-Speciality.

NULL values in the dataset column 'workclass' and 'occupation' are replaced by Unknown. They carried only about 5% of data and replacing with mode would increase the count of mode value, leading to biased behavior in ML algorithms.
